edx/xblock-utils
# [MCKIN-7023] Allows translated strings in XBlock Django templates

Adds `i18n` support to XBlock Django templates by adding support for `trans` and `blocktrans` template tags during template rendering.

The XBlock translations are provided by the `i18n` runtime service, which must be passed to `ResourceLoader.render_django_template`.

In order for an XBlock's template strings to be translated, the XBlock must:

* Require the `i18n` service, see e.g. [xblock-drag-and-drop-v2](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/26a8c5c5edb65beae24707353432654b50e7a230/drag_and_drop_v2/drag_and_drop_v2.py#L34)
   ```python
   @XBlock.needs('i18n')
   class MyXBlock(XBlock):
   ...
   ```
* Render templates using [`ResourceLoader.render_django_template`](https://github.com/open-craft/xblock-utils/blob/MCKIN-7023-template-i18n/xblockutils/resources.py#L50), passing the `i18n` service,  e.g.:
   ```python
   def studio_view(self, context):
      fragment = Fragment()
      fragment.add_content(loader.render_django_template('/templates/html/studio_edit.html', 
                                                         i18n_service=self.runtime.service(self, 'i18n'))
   ```
* In the HTML template, [load the `i18n` template tags](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/dd91427874637b8abad7fdeacaf45dfb50afaee1/drag_and_drop_v2/templates/html/drag_and_drop_edit.html#L1):
   ```html
   {% load i18n %}
   ```
* Mark translatable strings using the [`trans`](https://docs.djangoproject.com/en/2.0/topics/i18n/translation/#trans-template-tag) and [`blocktrans`](https://docs.djangoproject.com/en/2.0/topics/i18n/translation/#blocktrans-template-tag) tags, as described in the Django documentation.
* Extract and generate the translation `.po` and `.mo` files in your XBlock's `translations` directory, as described in the [XBlock tutorial](http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/edx_platform/edx_lms.html#internationalization-support).

**JIRA tickets**: [MCKIN-7023](https://edx-wiki.atlassian.net/browse/MCKIN-7023), [OSPR-2279](https://openedx.atlassian.net/browse/OSPR-2279)

**Discussions**:
* [WL-230](https://openedx.atlassian.net/browse/WL-230)
* [YONK-879](https://openedx.atlassian.net/browse/YONK-879?focusedCommentId=307771&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-307771)
* [edx-code mailing list](https://groups.google.com/forum/#!searchin/edx-code/xblock$20i18n%7Csort:date/edx-code/XoIQwvIMi4c/rwlyk76aAwAJ)
* [openedx-translation mailing list]( https://groups.google.com/d/msg/openedx-translation/cLR5tZI5oqQ/Eln357QzAwAJ)

**Dependency of:**
* https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/154
* https://github.com/edx-solutions/xblock-image-explorer/pull/37
* https://github.com/edx/edx-platform/pull/17670

**Screenshots**: See https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/154 for before and after screenshots using dummy translations.

**Sandbox URL**:  sandbox is being provisioned.

* LMS: https://pr17670.sandbox.opencraft.hosting/
* Studio: https://studio-pr17670.sandbox.opencraft.hosting/

**Merge deadline**: ASAP

**Testing instructions**:

See https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/154 for test instructions.

**Reviewers**
